### PR TITLE
Isaac/retention policy cron expression

### PR DESCRIPTION
**Pull request by IsaacCalligeros95**
Adding Retention Policy Configuration with Cron Expression. See https://github.com/OctopusDeploy/OctopusDeploy/pull/18021 for details.

[sc-40132]

### CODE REVIEW

File: source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+namespace Octopus.Client.Model
+{
+    public class RetentionPoliciesConfigurationResource : IResource
+    {
+        public RetentionPoliciesConfigurationResource()
+        {
+            Id = "retentionpolicies";
+        }
+
+        public string Id { get; set; }
```
Comment (veochen-octopus) [inline source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs:13]: Maybe dumb question, why is the `Id` on a configuration resource? Does this mean we can set the id?

Comment (IsaacCalligeros95) [reply source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs:13]: I'm not exactly sure but we store Configuration documents as ID, json pairs.
image
These are all defined in server with the base class ConfigurationDocument which takes the ID as a part of the constructor. I assume the Id is just there for matching purposes and can't or at least shouldn't be changed.

Comment (benPearce1) [reply source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs:13]: This should probably inherit from `Resource` which already contains an Id property.
If I remember correctly there should be no POST to create configuration records, so you can't just set any Id and POST it to the API, should require a GET and PUT.

Comment (IsaacCalligeros95) [reply source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs]: > This should probably inherit from `Resource` which already contains an Id property. If I remember correctly there should be no POST to create configuration records, so you can't just set any Id and POST it to the API, should require a GET and PUT.

The implementations are a bit all over the shop, some like ArchiveLimitConfigurationResource and EventRetentionConfigurationResource use IResource. These we're the most recent implementations of ConfigurationResources so I went with that approach. I've updated to use Resource and left the others as is, let me know you think it's worth updating those too.

Comment (benPearce1) [reply source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs]: If you change them, does it change the surface area test?

Comment (IsaacCalligeros95) [reply source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs]: Yep, 
image

Comment (benPearce1) [reply source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs]: Then just leave it as is

**Review by benPearce1 — APPROVED** [review]

File: source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs (hunk 1 above)
Comment (veochen-octopus) [reply source/Octopus.Server.Client/Model/RetentionPoliciesConfigurationResource.cs]: Only reason I ask is that configuration's ID is a db level implementation detail and is used to map between a db row and its correct configuration type, i.e. the same id always maps to the same type and vice verse. Apart from the document store nothing else should need to know about the id. So IMO exposing the id in the resource is leaky. But seems like we don't have any strict rules around this. Just raising for future consideration.

**Pull request merged**